packages/next/server/render.tsx

```diff
@@ -1798,8 +1807,14 @@ function renderToStream({
           reject(err)
```

Comment:
Hrm, would be nice to get rid of this (and the outer manually constructed promise). Maybe instead we could do something like this:

```tsx
let completionCallback: (error?: Error) => void
const completedWithoutErrors: Promise<void> = new Promise((resolve, reject) => {
  let resolved = false
  completionCallback = (error?: Error) => {
    if (!resolved) {
      resolved = true
      if (error) {
        reject(error}
      } else {
        resolve()
      }
    }
  }
})
```

and then call `completionCallback(error)` in `onError`, and `completionCallback()` in `onCompleteAll`. That way we'll still pass any errors that occur up and reject the overall promise if we `await completedWithoutErrors` and an error occurs.

Reply:
In other words:

```tsx
async function renderToStream({
  ReactDOMServer,
  element,
  suffix,
  dataStream,
  generateStaticHTML,
  flushEffectHandler,
}: {
  ReactDOMServer: typeof import('react-dom/server')
  element: React.ReactElement
  suffix?: string
  dataStream?: ReadableStream<Uint8Array>
  generateStaticHTML: boolean
  flushEffectHandler?: () => Promise<string>
}): Promise<ReadableStream<Uint8Array>> {
  let completionCallback: (error?: Error) => void
  const completedWithoutErrors: Promise<void> = new Promise((resolve, reject) => {
    let resolved = false
    completionCallback = (error?: Error) => {
      if (!resolved) {
        resolved = true
        if (error) {
          reject(error)
        } else {
          resolve()
        }
      }
    }
  })

  const renderStream: ReadableStream<Uint8Array> = await (
    ReactDOMServer as any
  ).renderToReadableStream(element, {
    onError(err: Error) {
      completionCallback(err)
    },
    onCompleteAll() {
      completionCallback()
    },
  })

  if (generateStaticHTML) {
    await completedWithoutErrors
  }

  const closeTag = '</body></html>'
  const suffixUnclosed = suffix ? suffix.split(closeTag)[0] : null
  const transforms: Array<TransformStream<Uint8Array, Uint8Array>> = [
      createBufferedTransformStream(),
      flushEffectHandler
        ? createFlushEffectStream(flushEffectHandler)
        : null,
      suffixUnclosed != null
        ? createPrefixStream(suffixUnclosed)
        : null,
      dataStream ? createInlineDataStream(dataStream) : null,
      suffixUnclosed != null ? createSuffixStream(closeTag) : null,
    ].filter(Boolean) as any

    return transforms.reduce(
      (readable, transform) => pipeThrough(readable, transform),
      renderStream
    )
}
```